torchao/prototype/quantization/codebook/codebook_ops.py

```diff
@@ -0,0 +1,317 @@
+        # don't need to compute input_chunk squared norm as it's constant during argmax
+
+        codes[chunk_start:chunk_end] = distances.argmax(dim=1)
+
+    codes = codes.view(num_out_blocks, num_in_blocks)
+
+    return codes.to(code_dtype)
+
+
+def dequantize_codebook(
+    codes: torch.Tensor,
+    codebook: torch.Tensor,
+    scales: torch.Tensor,
+    output_dtype: torch.dtype = torch.float32,
+) -> torch.Tensor:
+    """
+    Reconstructs the original tensor from codes and the codebook.
+
+    Args:
+        codes (torch.Tensor): Indices of codebook entries for each block,
+                                          shape (num_out_blocks, num_in_blocks).
+        codebook (torch.Tensor): Codebook tensor used for quantization,
+                                 shape (k, out_block_size, in_block_size).
+        output_dtype (torch.dtype): dtype for the output tensor.
+
+    Returns:
+        dequant (torch.Tensor): Reconstructed tensor, shape (out_features, in_features), where
+                      out_features = num_out_blocks * out_block_size and in_features = num_in_blocks * in_block_size.
+    """
+    assert output_dtype in [
+        torch.float32,
+        torch.float16,
+        torch.bfloat16,
+    ], f"Unsupported output dtype: {output_dtype}"
+
+    _, out_block_size, in_block_size = codebook.shape
+    num_out_blocks, num_in_blocks = codes.shape
+
+    # Use codes to lookup corresponding codebook entries and reshape
+    dequant = codebook[
+        codes
+    ]  # shape: [num_out_blocks, num_in_blocks, out_block_size, in_block_size]
+
+    dequant = dequant.permute(0, 2, 1, 3).reshape(
+        num_out_blocks * out_block_size, num_in_blocks * in_block_size
+    )
+
+    dequant = dequant.view(num_out_blocks * out_block_size, scales.shape[1], -1)
+    dequant = dequant * scales
+
+    dequant = dequant.view(num_out_blocks * out_block_size, -1)
+
+    return dequant.to(output_dtype)
+
+
+def choose_qparams_codebook(
+    input_tensor: torch.Tensor,
+    block_size: Tuple[int, ...],
+    scale_block_size: int,
+    code_dtype: torch.dtype,
+    max_iter: int = 200,
+    devices: Optional[List[torch.device]] = None,
+) -> torch.Tensor:
+    """
+    Initialize the codebook using k-means clustering on blocks of the input tensor.
+
+    Args:
+        input_tensor (torch.Tensor): The input tensor to be quantized.
+        block_size (Tuple[int, ...],): The size of the blocks for k-means clustering.
+        code_dtype (torch.dtype): The dtype for the codes.
+        max_iter (int): Maximum number of k-means iterations.
+        devices (List[torch.device]): Devices to run k-means on.
+
+    Returns:
+        torch.Tensor: The codebook tensor, shape (codebook_size, block_height, block_width).
+    """
+    if code_dtype == torch.int32:
+        codebook_size = 2**16
+    else:
+        codebook_size = _DTYPE_TO_QVALUE_BOUNDS[code_dtype][1] + 1
+
+    out_block_size, in_block_size = block_size
```

Comment:
I made it 2d because that is how it was implemented in AQLM code. The code will be a little more complicated, but it should be possible to generalize to more than 2d.